Native OSK

kd-11 opened the pull request:
- Implements a native osk input interface for use with a controller. Only ascii input in supported at the moment.
- Fixes a few bugs with the backend rendering of some native elements
- Improves font and text rendering system to support more glyphs and support more languages

Megamouse commented:
I'll check my personal tests here:

Games:
- [ ] Diablo 3
- [ ] Nascar '15
- [ ] Class of Heros 2G
- [ ] Nier
- [ ] Minecraft
- [ ] Skate 3

Flags:
- [ ] No Space
- [ ] Multiline

Megamouse commented on rpcs3/Emu/Cell/Modules/cellOskDialog.cpp:
```diff
@@ -85,7 +115,7 @@ error_code cellOskDialogLoadAsync(u32 container, vm::ptr<CellOskDialogParam> dia
 	bool result = false;
 
-	osk->on_close = [maxLength, wptr = std::weak_ptr<OskDialogBase>(osk)](s32 status)
+	osk->on_osk_close = [maxLength, wptr = std::weak_ptr<OskDialogBase>(osk)](s32 status)
```
the native osk has a chance to commit sudoku in the while loop containing a yield inside the osk_confirm_callback section

Megamouse commented on rpcs3/Emu/RSX/Overlays/overlay_osk.cpp:
```diff
@@ -0,0 +1,563 @@
+		void osk_dialog::on_enter(const std::string&)
+		{
+			// Accept dialog
+			Close(true);
```
```suggestion
			if (!(flags & CELL_OSKDIALOG_NO_RETURN)/* && !(flags & CELL_OSKDIALOG_NO_SPACE)*/)
			{
				on_default_callback("\n");
			}
			else
			{
				// Beep or give some other kind of visual feedback
			}
```

Megamouse commented on rpcs3/Emu/RSX/Overlays/overlay_osk.cpp:
```diff
@@ -0,0 +1,563 @@
+
+			m_preview.set_pos(frame_x, frame_y + 30);
+			m_preview.set_size(frame_w, 40);
+			m_preview.set_text(initial_text.empty()? "[Enter Text]" : initial_text);
```
Maybe also set the default preview when the Text field is empty (up for debate)

kd-11 replied:
Hint text color changed. If needed everytime the field is empty, its easy to add again.

MSuih commented:
Would it be possible to have backspace bound to some button? It's cubersome to move the cursor to that button every time one wants to erase something. Space as a button woudn't be too bad of an idea either. In addition I think the PS3 OSK also allows you to move the caret around, you could consider implementing that with shoulder buttons as well.

Also I don't like having shift and accept/cancel as face buttons. A more natural spot for shift would be a shoulder button imo and start would be more fitting as an accept button. I'm not sure what to do with cancel though, maybe it could be placed on the keyboard with no button assigned to it.

kd-11 commented:
I don't have access to all buttons in native dialog right now without restructuring how pad input works. Caret itself doesn't exist yet as well and many other features. Right now its more important to just check that games do not crash.

kd-11 commented:
Added more buttons and remapped them to more closely match ps3. I think only the edit_text control remains, which will be important to support multiline and caret positioning.

MarioSonic2987 commented:
If you wait enough time (like ~5 seconds), Worms Ultimate Mayhem crashes after pressing Cancel or Accept. In master and this PR without enabling native user interface, works fine.

Master:
![image](https://user-images.githubusercontent.com/20027681/51990372-2deb9e80-24a9-11e9-8a65-fe1743a53b79.png)
![image](https://user-images.githubusercontent.com/20027681/51990049-82dae500-24a8-11e9-86fe-90742c29add1.png)
[Log WUM 30-1-19 Master.gz](https://github.com/RPCS3/rpcs3/files/2813129/Log.WUM.30-1-19.Master.gz)
[Log WUM OSK.zip](https://github.com/RPCS3/rpcs3/files/2813116/Log.WUM.30-1-19.OSK.VK.zip)

kd-11 commented:
Added a proper edit control with support for multiline text if requested by the application. No scrolling available so you're limited to 3 lines max which should be good enough. Also adds a caret for visual feedback when space is pressed for example.
Also added support for disabled keys e.g disabled enter key when multiline is disabled or spacebar when input with no spaces is requested.
If all goes well, I'll go ahead with the config option tomorrow.

kd-11 commented:
How to cause the crash? I cannot get it to crash (tested with DeS)

kd-11 commented:
@MSuih Fixed now. The crashes are possibly because the osk dialog never truly closes, just resets the interface. I must have failed to reset some variable.

kd-11 commented:
Identified a bug in the overlay renderer where resizing the window from 1280x720 breaks clip regions. This must have silently affected save dialog as well (I suspect some entries were cut off but it was not reported). Will push a fix asap.

mavethee commented:
@kd-11 When I wanna erase space in default name of world, game crashes...

mavethee commented:
And there is an isssue with special letters, but probably game crash is not caused by that.

kd-11 commented:
@mavethee What is the default world name?
As for special characters, I did mention only ascii input is supported. Same as all native dialog elements. Needs a lot of font work before non-standard text is supported.

MSuih commented:
I think the issue he's having is that the default text contains non-ascii characters and your implementation assumes that only ascii characters have been entered. Maybe the backspace function uses a fixed chararter size or something? If you need a bandaid solution then wiping all non-ascii characters from default text could work.

kd-11 commented:
Backspace just does text.substr(), not sure how that would cause the crashing, but the out of range definately hints at something of the sort where the special character is larger than ascii limit of 128. I'll run some tests with some non-ascii characters and see if I can reproduce.

mavethee commented:
![image](https://user-images.githubusercontent.com/47083527/52052951-4a92df80-2558-11e9-8eb7-944fdbc0c0bf.png)
"Nowy Świat" (in polish "New World").
I can't erase the space between "Nowy" and "Świat", because of this error.

kd-11 commented:
Had to just strip all extended codes and replace with the character '#'. Fixes crashing on polish characters. If the '#' replacement is too jarring, I'm open to alternative placeholder characters. Note that the replaced character is 'real' in that it exists in the string if you choose to confirm it. Its not just a visual effect/trick.

mavethee commented:
@kd-11 Works well, but can you replace '#' with normal letters?
I mean, ś to s, ó - o, ź - z, ż - z, ć - c, ą - a, ę - e.
Of course, if you want to and if it's possible. My programming skills are still bacics of basics :)

MSuih commented:
There are too many different variations for that to be feasible. For example there's over 30 different variations of letter A alone.

mavethee commented:
Ouch. So only replacement with "#" left.  But thanks for fixing that error with erasing space between words :)

kd-11 commented:
Too many variations unfortunately. Its really easy to write a replacement table manually but thats a near impossible task given how many variations there are. I plan to eventually replace all text in the native UI with proper unicode support which will remove this limitation (all text variables encoded in utf-32). This is not going to happen any time soon though.

mavethee commented:
@kd-11 Nice. I will be waiting and learn some more stuff, because I wanna help you guys. Going to IT school in Sempteber, but learning on my own it's probably better :P

kd-11 commented:
I'll hackishly map downward the characters instead of truly converting utf-16 to utf-8. Should help preserve characters in the high ascii range.

kd-11 commented:
Upgraded the text rendering system to use extended ascii with all text converted to this representation. This allows a lot more glyphs to render correctly, and (e.g) makes french text readable. Not all characters can be represented yet, but its at least a step to support more languages.

MarioSonic2987 commented:
Spanish texts are now readable (now renders glyphs like **áéíóúñü**):
![image](https://user-images.githubusercontent.com/20027681/52075455-9cedf380-258c-11e9-865d-898894768624.png)
![image](https://user-images.githubusercontent.com/20027681/52075591-f3f3c880-258c-11e9-90c7-29213592cc7c.png)
![image](https://user-images.githubusercontent.com/20027681/52075838-914efc80-258d-11e9-9379-41d477e3c9aa.png)

mavethee commented:
![image](https://user-images.githubusercontent.com/47083527/52079419-d119e200-2595-11e9-9cad-3feca677f119.png)
Almoust nice! RPCS3 will be better than PS3 someday :)

leskal commented:
Greek letters in Sly Cooper Thieves in Time
Master:
![master](https://user-images.githubusercontent.com/11486824/52086027-8f496580-25ae-11e9-8913-42fa81611cb6.png)
PR:
![pr](https://user-images.githubusercontent.com/11486824/52086171-d9cae200-25ae-11e9-8b36-f0491b21e935.png)
